fix/feat: Move convolution core to `impl` + add feature (FX converter refactor)

gs-olive opened the pull request:
# Description

- Centralize `convolution` implementation in FX across all source IR variants, including support for `conv1d`, quantized, and other configurations
- Update reference implementations across the stack to use centralized utility and remove individual replicated implementations
- Allow conv layers to take bias inputs in FX, per new functionality from TRT
- Enable pass-through of build errors in Dynamo e2e tests to ensure errors are not being hidden (this PR fixes a bug which disallowed that pass-through)

Fixes #1954
Addresses first bug in #1565 

## Type of change
- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)

# Checklist:

- [ x ] My code follows the style guidelines of this project (You can use the linters)
- [ x ] I have performed a self-review of my own code
- [ x ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ x ] I have made corresponding changes to the documentation
- [ x ] I have added tests to verify my fix or my feature
- [ x ] New and existing unit tests pass locally with my changes
- [ x ] I have added the relevant labels to my PR in so that relevant reviewers are notified

gs-olive commented on py/torch_tensorrt/fx/converters/impl/convolution.py:
```diff
@@ -0,0 +1,146 @@
+            network, target, tuple(), kwargs, name + "_unsqueeze"
+        )
+
+    # Process bias terms
+    if isinstance(bias, torch.Tensor):
+        # Transform the bias constant into a Numpy array
+        bias = to_numpy(bias)
+
+    elif isinstance(bias, TRTTensor):
+        bias = get_trt_tensor(network, bias, f"{name}_bias")
+
```
Did not add an unsqueeze operation to bias term since the requirement in TRT for the bias term is that it must have number of elements equal to the number of output features of the convolution, so the same bias as is used for Conv1D would work for Conv2D, with the number of output features being fixed.

frank-wei replied:
just to clarify, you mean 1D or 2D conv? For 1D, we need bias to be unsqueezed.

gs-olive replied:
I initially meant for all conv layers since [this documentation](https://docs.nvidia.com/deeplearning/tensorrt/api/python_api/infer/Graph/Layers.html#iconvolutionlayer) seems to indicate we just need the number of elements in the bias Tensor to the be correct, and not necessarily the dimensions, but if bias needs to be unsqueezed for 1D, I can add that functionality back. I am wondering if the intended unsqueeze should be in the first dimension (`torch.unsqueeze(bias, 0)`) or the last dimension (`torch.unsqueeze(bias, -1)`)?

frank-wei replied:
I am wondering if TRT has done the broadcast internally since the unit test for 1D works good even though you did not unsqueeze it.

gs-olive replied:
I verified on a small sample that Conv1D with bias compiles + runs inference successfully without unsqueezing the bias term

narendasan replied:
Should we have a unit test for this just to catch if TRT behavior changes?

gs-olive replied:
I could add one, but it would likely be very similar to this case, which interprets, builds, and runs inference with a Conv1D model with TRT both with and without bias:
https://github.com/pytorch/TensorRT/blob/cae6b7c694db1b6a5f287f9fc4e6c40ba08e1079/py/torch_tensorrt/fx/test/converters/acc_op/test_convolution.py#L8-L45

narendasan commented on py/torch_tensorrt/fx/converters/acc_ops_converters.py:
```diff
@@ -268,32 +159,36 @@ def acc_ops_conv_transposend(
-            f"ConvTranspose {name} has bias of type {type(kwargs['bias'])}, Expect Optional[Tensor]"
-        )
-    bias = to_numpy(kwargs["bias"])  # type: ignore[arg-type]
+    if not isinstance(kwargs["bias"], TRTTensor):
```
Do this in a separate PR and centralize conv^T

gs-olive replied:
Instanced #1996 for this

narendasan commented on py/torch_tensorrt/fx/converters/convolution.py:
```diff
@@ -1,212 +1,123 @@
 
 
 @tensorrt_converter(torch.nn.modules.conv.Conv1d)
 def conv1d(network, submod, args, kwargs, layer_name):
```
Move this to nn_converters

narendasan commented on py/torch_tensorrt/fx/converters/convolution.py:
```diff
@@ -1,212 +1,123 @@
 
 
 @tensorrt_converter(torch.nn.modules.conv.Conv2d)
 def conv2d(network, submod, args, kwargs, layer_name):
```
Same here, we should get rid of this file and just put the registrations for each opset in the same file

narendasan reviewed (commented):
Mostly organization stuff

gs-olive commented on py/torch_tensorrt/fx/converters/nn_ops_converters.py:
Move `convolution.py` contents to `nn_ops_converters.py` for code cleanliness + organization. Remove unused imports.

gs-olive commented:
HI @wushirong - thank you for the review. I was wondering if you could have another look at the changes, as I've moved `convolution.py` contents to `nn_ops_converters.py` for code cleanliness + organization and removed unused imports.
